python/hail/tests/test_typecheck2.py

```diff
@@ -0,0 +1,362 @@
+from hail.typecheck2 import typecheck
+import unittest
+from typing import *
+
+
+def f_noargs():
+    pass
+
+
+def f_simple1(x: int):
+    typecheck(f_simple1)
+
+
+def f_simple2(x: str):
+    typecheck(f_simple2)
+
+
+def f_optional(x: Optional[str]):
+    typecheck(f_optional)
+
+
+def f_float(x: float):
+    typecheck(f_float)
+
+
+def f_union(x: Union[int, str]):
+    typecheck(f_union)
+
+
+def f_list(x: List):
+    typecheck(f_list)
+
+
+def f_list_int(x: List[int]):
+    typecheck(f_list_int)
+
+
+def f_sequence(x: Sequence):
+    typecheck(f_sequence)
+
+
+def f_sequence_float(x: Sequence[float]):
+    typecheck(f_sequence_float)
+
+
+def f_set(x: Set):
+    typecheck(f_set)
+
+
+def f_set_union(x: Set[Union[int, str]]):
+    typecheck(f_set_union)
+
+
+def f_frozenset(x: FrozenSet):
+    typecheck(f_frozenset)
+
+
+def f_frozenset_int(x: FrozenSet[int]):
+    typecheck(f_frozenset_int)
+
+
+def f_collection(x: Collection):
+    typecheck(f_collection)
+
+
+def f_collection_str(x: Collection[str]):
+    typecheck(f_collection_str)
+
+
+def f_tuple(x: Tuple):
+    typecheck(f_tuple)
+
+
+def f_tuple2(x: Tuple[Any, ...]):
+    typecheck(f_tuple2)
+
+
+def f_tuple_var(x: Tuple[int, ...]):
+    typecheck(f_tuple_var)
+
+
+def f_tuple_sized1(x: Tuple[int]):
+    typecheck(f_tuple_sized1)
+
+
+def f_tuple_sized2(x: Tuple[int, str, Union[str, int]]):
+    typecheck(f_tuple_sized2)
+
+
+def f_dict(x: Dict):
+    typecheck(f_dict)
+
+
+def f_dict_str_int(x: Dict[str, int]):
+    typecheck(f_dict_str_int)
+
+
+def f_mapping(x: Mapping):
+    typecheck(f_mapping)
+
+
+def f_mapping_str_int(x: Mapping[str, int]):
+    typecheck(f_mapping_str_int)
+
+
+def f_varargs(x: int, *xs: str):
+    typecheck(f_varargs)
+
+
+def f_varargs2(*xs: str):
+    typecheck(f_varargs2)
+
+
+def f_varargs3(x: int, *xs: Union[str, int]):
+    typecheck(f_varargs3)
+
+
+def f_kwargs(x: int, **xs: int):
+    typecheck(f_kwargs)
+
+
+def f_kwargs2(x: int, **xs: List[Union[str, int]]):
+    typecheck(f_kwargs2)
+
+
+def f_complicated(x: List[List[Set[Union[str, int]]]],
+                  y: Dict[Union[int, str], Sequence[Optional[int]]]):
+    typecheck(f_complicated)
+
+def f_callable(x: Callable[[int, str], str]):
+    typecheck(f_callable)
+
+
+class Tests(unittest.TestCase):
```

Comment:
The organization of the tests makes it difficult to tell by inspection how each type checker is being exercised. I think it would be clearer if the tests were grouped by type checker, e.g.:
```
def test_check_frozenset(self):
    def f_frozenset(x: FrozenSet):
        typecheck(f_frozenset)

    def f_frozenset_int(x: FrozenSet[int]):
       typecheck(f_frozenset_int)

    # should not throw
    f_frozenset(frozenset())
    f_frozenset(frozenset([1, 2, 3]))
    f_frozenset_int(frozenset())
    f_frozenset_int(frozenset([1, 2, 3]))

    # should throw
    with self.assertRaises(TypeError):
        f_frozenset(set())
    with self.assertRaises(TypeError):
        f_frozenset([])
    with self.assertRaises(TypeError):
        f_frozenset_int(set())
    with self.assertRaises(TypeError):
        f_frozenset_int(frozenset(['1']))
```

This makes it much easier to compare the implementation of `check_frozenset` with the tests that verify it.

Reply:
somewhat painstaking to fix, but you're right that the reorg made it a LOT cleaner.